Sources/PackageModel/ManifestSourceGeneration.swift

```diff
@@ -0,0 +1,326 @@
+                params.append(SourceCodeFragment("condition: \(configNode.generateSourceCode())"))
+                self.init(".product", delimiters: .parentheses, multiline: false, subnodes: params)
+            }
+            else {
+                self.init(name.quotedForPackageManifest)
+            }
+        }
+    }
+    
+    /// Instantiates a SourceCodeFragment to represent a single dependency condition.
+    convenience init(from condition: PackageConditionDescription) {
+        var params: [SourceCodeFragment] = []
+
+        if !condition.platformNames.isEmpty {
+            let nodes = condition.platformNames.map{ SourceCodeFragment(".\($0)") }
+            params.append(SourceCodeFragment(key: "platforms", multiline: false, subnodes: nodes))
+        }
+        if let configName = condition.config {
+            params.append(SourceCodeFragment(key: "configuration", enum: configName))
+        }
+        self.init(".when", delimiters: .parentheses, multiline: false, subnodes: params)
+    }
+
+    /// Instantiates a SourceCodeFragment to represent a single Swift language version.
+    convenience init(from version: SwiftLanguageVersion) {
+        let version = String(describing: version)  // FIXME
+        self.init(".version(\(version.quotedForPackageManifest))")
+    }
+}
+
+
+/// Convenience initializers for key-value pairs of simple types.
+fileprivate extension SourceCodeFragment {
+    
+    /// Initializes a SourceCodeFragment for a boolean in a generated manifest.
+    convenience init(key: String, boolean: Bool) {
+        self.init(key + ": " + (boolean ? "true" : "false"))
+    }
+
+    /// Initializes a SourceCodeFragment for a quoted string in a generated manifest.
+    convenience init(key: String, string: String) {
+        self.init(key + ": " + string.quotedForPackageManifest)
+    }
+
+    /// Initializes a SourceCodeFragment for an enum in a generated manifest.
+    convenience init(key: String, enum: String) {
+        self.init(key + ": ." + `enum`)
+    }
+
+    /// Initializes a SourceCodeFragment for a string list in a generated manifest.
+    convenience init(key: String, strings: [String]) {
+        let subnodes = strings.map{ SourceCodeFragment($0.quotedForPackageManifest) }
+        self.init(key + ": ", delimiters: .brackets, subnodes: subnodes)
+    }
+
+    /// Initializes a SourceCodeFragment for a list of nodes in a generated manifest.
+    convenience init(key: String, multiline: Bool = true, subnodes: [SourceCodeFragment]) {
+        self.init(key + ": ", delimiters: .brackets, multiline: multiline, subnodes: subnodes)
+    }
+}
+
+
+
+/// Helper type to emit source code.  Represents one node of source code, as a
+/// arbitrary string followed by an optional child list, optionally enclosed in
+/// a pair of delimiters.  The code generation works by creating source code
+/// fragments and then rendering them as source code with proper formatting.
+fileprivate class SourceCodeFragment {
```

Comment:
is this actually carrying state and needs to be a class?

Reply:
I'd prefer a struct, but since it can have children of the same type, I got errors.  I think there's a way to have a by-reference struct, though, so I was going to make that change before making this no longer be a draft.

Reply:
I think `indirect` is what I might be looking for.  Will try that.